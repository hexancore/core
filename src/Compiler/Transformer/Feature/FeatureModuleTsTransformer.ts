import ts from 'typescript';
import type {FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import type { ProviderModuleMetaTransformDef } from '../ModuleClassTsTransformer';
import type { FeatureTransformContext } from './FeatureTransformContext';
import type { FeatureApplicationMessageMeta } from '../../../Util/Feature/Meta';
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";

/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureModuleTsTransformer extends AbstractFeatureTsTransformer {

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    return featureSourcePath.localSourcePath.endsWith(feature.name + "Module.ts");
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const feature = context.feature;

    const messageHandlersProviders: ProviderModuleMetaTransformDef[] = [];
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.commands));
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.queries));

    return this.moduleClassTransformer.transform({
      imports: [],
      meta: {
        imports: [],
        providers: [...messageHandlersProviders, ...this.createServiceProviders(feature)],
      },
      source,
      context: context.tsContext,
    });
  }

  private createMessageHandlerProviders(messages: FeatureApplicationMessageMeta[]): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const m of messages) {
      const importPath = `./${m.path}/${m.handlerClass}`;
      providers.push({
        addToExports: false,
        name: m.handlerClass,
        importFrom: importPath
      });
    }

    return providers;
  }

  private createServiceProviders(feature: FeatureMeta): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const s of feature.application.services) {
      if (!s.isInjectable) {
        continue;
      }

      providers.push({
        addToExports: false,
        name: s.name,
        importFrom: `./${s.path}`
      });
    }

    return providers;
  }
}