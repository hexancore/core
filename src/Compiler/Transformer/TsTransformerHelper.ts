import { Module } from "@nestjs/common";
import ts, { type TransformationContext } from "typescript";

export class TsTransfromerHelper {

  public static getLastImportIndex(source: ts.SourceFile): number {
    let lastIndex: number = 0;

    for (let i = 0; i < source.statements.length; ++i) {
      if (ts.isImportDeclaration(source.statements[i])) {
        lastIndex = i;
        continue;
      }

      if (ts.isDecorator(source.statements[i])) {
        break;
      }
    }

    return lastIndex;
  }

  public static transformDecoratorCallOfType(
    node: ts.Node,
    decoratorType: (...args) => any,
    updater: (decorator: ts.Decorator, decoratorExpression: ts.CallExpression) => ts.Node
  ): ts.Node {
    if (ts.isDecorator(node)) {
      const decoratorExpression = node.expression;
      if (ts.isCallExpression(decoratorExpression) && ts.isIdentifier(decoratorExpression.expression)) {
        if (decoratorExpression.expression.text === decoratorType.name) {
          return updater(node, decoratorExpression);
        }
      }
    }

    return node;
  }

  public static transformModuleDecoratorExpression(node: ts.Node, moduleMetaProvider: () => ({
    providers?: ts.Expression[],
    imports?: ts.Expression[],
    exports?: ts.Expression[],
  })): ts.Node {
    return TsTransfromerHelper.transformDecoratorCallOfType(node, Module, (decorator, decoratorExpression) => {
      const decoratorArguments = decoratorExpression.arguments;
      if (decoratorArguments.length > 0 && ts.isObjectLiteralExpression(decoratorArguments[0])) {
        const moduleMetaExpression = decoratorArguments[0];

        const assignedProps: string[] = [];
        const moduleMeta = moduleMetaProvider();
        const newProperties = moduleMetaExpression.properties.map(prop => {
          return TsTransfromerHelper.updatePropertyArrayAssigment(prop, (name: string) => {
            if (moduleMeta[name]) {
              assignedProps.push(name);
              return moduleMeta[name];
            }
            return null;
          });
        });

        for (const p in moduleMeta) {
          if (!assignedProps.includes(p)) {
            newProperties.push(ts.factory.createPropertyAssignment(p, ts.factory.createArrayLiteralExpression(moduleMeta[p])));
          }
        }
        const updatedModuleMetaExpression = ts.factory.updateObjectLiteralExpression(moduleMetaExpression, newProperties);
        const updatedDecoratorExpression = ts.factory.updateCallExpression(
          decoratorExpression,
          decoratorExpression.expression,
          decoratorExpression.typeArguments,
          [updatedModuleMetaExpression]
        );
        return ts.factory.updateDecorator(decorator, updatedDecoratorExpression);
      }

      return node;
    });
  }

  public static updatePropertyArrayAssigment(prop: ts.ObjectLiteralElementLike, newElementsProvider: (name: string) => ts.Expression[] | null): ts.ObjectLiteralElementLike {
    if (ts.isPropertyAssignment(prop) && ts.isIdentifier(prop.name)) {
      const newElements = newElementsProvider(prop.name.text);
      if (newElements) {
        if (ts.isArrayLiteralExpression(prop.initializer)) {
          const elements = [...prop.initializer.elements, ...newElements];
          return ts.factory.updatePropertyAssignment(prop, prop.name, ts.factory.updateArrayLiteralExpression(prop.initializer, elements));
        }
      }
    }

    return prop;
  }

  public static extractModuleSpecifierFromImportDeclaration(node: ts.ImportDeclaration): string {
    return (node.moduleSpecifier as ts.StringLiteral).text;
  }

  public static createNamedImportAccess(importDec: ts.ImportDeclaration, name: string, context: TransformationContext): ts.PropertyAccessChain | ts.Identifier {
    const opts = context.getCompilerOptions();

    if (opts.module === ts.ModuleKind.Node16) {
      return ts.factory.createPropertyAccessChain(ts.factory.getGeneratedNameForNode(importDec), undefined, name);
    }

    return ts.factory.createIdentifier(name);

  }

}