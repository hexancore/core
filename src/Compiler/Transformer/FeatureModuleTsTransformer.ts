import { execFileSync, execSync } from "node:child_process";
import ts from 'typescript';
import { type FeatureApplicationMessageInfo, type FeatureModuleDiscovery } from "../../Util/FeatureModuleDiscoverer";
import { TsTransfromerHelper } from "./TsTransformerHelper";

/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureModuleTsTransformer {
  public constructor(private features: Map<string, FeatureModuleDiscovery>) {
  }

  public static create(sourceRoot: string, compilerDir: string): FeatureModuleTsTransformer {
    const features = this.loadFeaturesMapSync(sourceRoot, compilerDir);
    return new this(features);
  }

  private static loadFeaturesMapSync(sourceRoot: string, compilerDir: string): Map<string, FeatureModuleDiscovery> {
    // TS transformer can't use async, but we need async code and execSync allows run async code :) Hack of day !
    const output = execFileSync('node', [`${compilerDir}/discoverer.js`, sourceRoot]);
    const features = JSON.parse(output.toString());
    return new Map(features);
  }

  public transform(featureName: string, source: ts.SourceFile, context: ts.TransformationContext): ts.SourceFile {
    const feature = this.features.get(featureName);
    if (!feature) {
      throw new Error(`No found feature module info, name: ${featureName}`);
    }

    let imported = false;
    const currentImports = new Set<string>();
    let importMap: Map<string, ts.ImportDeclaration>;

    const visitor: ts.Visitor = (node) => {
      if (ts.isImportDeclaration(node)) {
        currentImports.add(TsTransfromerHelper.extractModuleSpecifierFromImportDeclaration(node));
        return node;
      }

      if (!ts.isImportDeclaration(node) && !imported) {
        imported = true;
        importMap = this.createImportDeclarations(feature, currentImports);
        if (ts.isClassDeclaration(node)) {
          node = ts.visitEachChild(node, (n) => this.transformModuleDecorator(n, feature, importMap, context), context);
        }
        return [...importMap.values(), node];
      }

      if (ts.isClassDeclaration(node)) {
        node = ts.visitEachChild(node, (n) => this.transformModuleDecorator(n, feature, importMap, context), context);
      }

      return node;

    };

    return ts.factory.updateSourceFile(source, ts.visitNodes(
      source.statements,
      visitor,
      (node): node is ts.Statement => ts.isStatement(node))
    );
  }

  private createImportDeclarations(feature: FeatureModuleDiscovery, currentImports: Set<string>) {
    const importMap: Map<string, ts.ImportDeclaration> = new Map();

    this.addMessageHandlerImportDeclarations(feature.application.commands, currentImports, importMap);
    this.addMessageHandlerImportDeclarations(feature.application.queries, currentImports, importMap);
    this.addMessageHandlerImportDeclarations(feature.application.events, currentImports, importMap);
    this.addServiceImportDeclarations(feature, currentImports, importMap);
    this.addInfrastructureModuleImportDeclaration(feature, importMap);

    return importMap;
  }

  private addMessageHandlerImportDeclarations(messages: FeatureApplicationMessageInfo[], currentImports: Set<string>, importMap: Map<string, ts.ImportDeclaration>) {
    for (const m of messages) {
      const importPath = `./${m.path}/${m.handlerClassName}`;
      if (currentImports.has(importPath)) {
        continue;
      }

      importMap.set(m.handlerClassName, this.createImportDeclaration(m.handlerClassName, importPath));
    }
  }

  private addServiceImportDeclarations(feature: FeatureModuleDiscovery, currentImports: Set<string>, importMap: Map<string, ts.ImportDeclaration>): void {
    for (const s of feature.application.services) {
      if (s.isInjectable) {
        const importPath = `./${s.path}`;
        if (currentImports.has(importPath)) {
          continue;
        }

        importMap.set(s.className, this.createImportDeclaration(s.className, importPath));
      }
    }
  }
  private addInfrastructureModuleImportDeclaration(feature: FeatureModuleDiscovery, importMap: Map<string, ts.ImportDeclaration>): void {
    const infrastructureModuleImportDec = this.createImportDeclaration(feature.infrastructure.module.className, `./${feature.infrastructure.module.path}`);
    importMap.set(feature.infrastructure.module.className, infrastructureModuleImportDec);
  }

  private createImportDeclaration(importName: string, importPath: string): ts.ImportDeclaration {
    const namedImports = ts.factory.createNamedImports([
      ts.factory.createImportSpecifier(false, undefined, ts.factory.createIdentifier(importName))
    ]);
    const importFrom = ts.factory.createStringLiteral(importPath);
    return ts.factory.createImportDeclaration(
      undefined,
      ts.factory.createImportClause(false, undefined, namedImports),
      importFrom
    );
  }

  private transformModuleDecorator(
    node: ts.Node,
    feature: FeatureModuleDiscovery,
    importMap: Map<string, ts.ImportDeclaration>,
    context: ts.TransformationContext
  ) {
    return TsTransfromerHelper.transformModuleDecoratorExpression(node, () => {
      const imports = [
        feature.infrastructure.module.className
      ];

      const providers = [
        ...feature.application.commands.map((m) => m.handlerClassName),
        ...feature.application.queries.map((m) => m.handlerClassName),
        ...feature.application.events.map((m) => m.handlerClassName),
        ...feature.application.services.filter((s) => s.isInjectable).map((s) => s.className)
      ];
      const classIdentifierMapper = (className) => TsTransfromerHelper.createNamedImportAccess(importMap.get(className)!, className, context);
      return {
        imports: imports.map(classIdentifierMapper),
        providers: providers.map(classIdentifierMapper),
      };
    });
  }
}