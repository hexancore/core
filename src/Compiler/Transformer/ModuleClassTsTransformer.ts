import ts from "typescript";
import { TsTransfromerHelper, type ImportedIdentifierMapper, type ImportFromMapper } from "./TsTransformerHelper";

export type ImportDeclarationMap = Map<string, ts.ImportDeclaration>;

export interface AddImportTransformDef {
  name: string;
  importModuleSpecifier: string;
}

export interface ImportModuleMetaTransformDef {
  name: string;
  importModuleSpecifier: string;
  addToExports: boolean;
}

export interface ProviderModuleMetaTransformDef {
  name: string;
  importFrom: string;
  expression?: ts.Expression;
  addToExports: boolean;
}

export interface CreateImportDeclarationsOutput {
  declarations: ImportDeclarationMap;
  newDefImports: ImportModuleMetaTransformDef[];
  newDefProviders: ProviderModuleMetaTransformDef[];
  providerStatements: ts.VariableStatement[];
  extraMetaProvider?: ts.Identifier;
}

export interface ModuleClassTsTransformOptions {
  imports?: AddImportTransformDef[],
  meta?: {
    imports: ImportModuleMetaTransformDef[];
    providers: ProviderModuleMetaTransformDef[];
  };
  extraStatementProvider?(importedIdentifierMapper: ImportedIdentifierMapper): ts.Statement[];
  extraMetaProvider?(): ts.Identifier;
  source: ts.SourceFile;
  context: ts.TransformationContext;
}

export class ModuleClassTsTransformer {

  protected importedIdentifierMapper: (importDec: ts.ImportDeclaration, name: string, context: ts.TransformationContext) => ts.PropertyAccessChain | ts.Identifier;

  public constructor(protected importFromMapper: ImportFromMapper, protected needFixImportAccess: boolean) {
    this.importedIdentifierMapper = this.needFixImportAccess ? (importDec, name, context) =>
      TsTransfromerHelper.createNamedImportAccess(importDec, name, context)
      : (_importDec, name, _context) => ts.factory.createIdentifier(name);

  }

  public transform(options: ModuleClassTsTransformOptions): ts.SourceFile {
    let imported = false;
    const currentImports = new Set<string>();
    let importDeclarations: CreateImportDeclarationsOutput;

    const visitor: ts.Visitor = (node) => {
      if (ts.isImportDeclaration(node)) {
        currentImports.add(TsTransfromerHelper.extractModuleSpecifierFromImportDeclaration(node));
        return node;
      }

      if (!imported) {
        imported = true;
        importDeclarations = this.createImportDeclarations(options, currentImports);

        if (ts.isClassDeclaration(node)) {
          node = ts.visitEachChild(node, (childNode) => this.transformModuleDecorator(childNode, importDeclarations, options.context), options.context);
        }
        const importedIdentifierMapper = (className: string) =>
          this.importedIdentifierMapper(importDeclarations.declarations.get(className)!, className, options.context);

        let extraStatements: ts.Statement[] = [];
        if (options.extraStatementProvider) {
          extraStatements = options.extraStatementProvider(importedIdentifierMapper);
        }

        return [...importDeclarations.declarations.values(), ...extraStatements, ...importDeclarations.providerStatements, node];
      }

      if (ts.isClassDeclaration(node)) {
        node = ts.visitEachChild(node, (childNode) => this.transformModuleDecorator(childNode, importDeclarations, options.context), options.context);
      }

      return node;

    };

    return ts.factory.updateSourceFile(options.source, ts.visitNodes(
      options.source.statements,
      visitor,
      (node): node is ts.Statement => ts.isStatement(node))
    );
  }

  private createImportDeclarations(options: ModuleClassTsTransformOptions, currentImports: Set<string>): CreateImportDeclarationsOutput {

    const declarations: ImportDeclarationMap = new Map();
    const newMetaImports: ImportModuleMetaTransformDef[] = [];
    const newMetaProviders: ProviderModuleMetaTransformDef[] = [];
    const providerStatements: ts.VariableStatement[] = [];

    if (options.imports) {
      for (const i of options.imports) {
        const importModuleSpecifier = this.importFromMapper(i.importModuleSpecifier);
        if (currentImports.has(importModuleSpecifier)) {
          continue;
        }

        declarations.set(i.name, TsTransfromerHelper.createImportDeclaration(i.name, importModuleSpecifier));
      }

    }

    if (options.meta) {
      for (const d of options.meta.imports) {
        const importModuleSpecifier = this.importFromMapper(d.importModuleSpecifier);
        if (currentImports.has(importModuleSpecifier)) {
          continue;
        }

        declarations.set(d.name, TsTransfromerHelper.createImportDeclaration(d.name, importModuleSpecifier));
        newMetaImports.push(d);
      }

      for (const d of options.meta.providers) {
        const importModuleSpecifier = this.importFromMapper(d.importFrom);
        if (currentImports.has(importModuleSpecifier)) {
          continue;
        }

        declarations.set(d.name, TsTransfromerHelper.createImportDeclaration(d.name, importModuleSpecifier));
        newMetaProviders.push(d);
        if (d.expression) {
          providerStatements.push(this.createProviderDeclarationStatement(d));
        }
      }
    }

    return {
      declarations,
      providerStatements,
      newDefImports: newMetaImports,
      newDefProviders: newMetaProviders,
      extraMetaProvider: options.extraMetaProvider ? options.extraMetaProvider() : undefined
    };
  }

  private transformModuleDecorator(
    node: ts.Node,
    importDeclarations: CreateImportDeclarationsOutput,
    context: ts.TransformationContext
  ) {
    return TsTransfromerHelper.transformModuleDecoratorExpression(node, () => {
      const imports = importDeclarations.newDefImports.map((i) => i.name);

      const classIdentifierMapper = (className: string) =>
        this.importedIdentifierMapper(importDeclarations.declarations.get(className)!, className, context);

      const providers: any[] = importDeclarations.newDefProviders.map(p => p.name).map(classIdentifierMapper);
      const exports: any[] = importDeclarations.newDefProviders.filter((p) => p.addToExports).map(p => p.name).map(classIdentifierMapper);

      if (importDeclarations.extraMetaProvider) {
        providers.push(this.getArrayPropertyAsSpread(importDeclarations.extraMetaProvider, 'providers'));
        exports.push(this.getArrayPropertyAsSpread(importDeclarations.extraMetaProvider, 'exports'));
      }
      return {
        imports: imports.map(classIdentifierMapper),
        providers,
        exports
      };
    });
  }

  private createProviderDeclarationStatement(def: ProviderModuleMetaTransformDef): ts.VariableStatement {
    const varName = def.name + "Provider";
    return TsTransfromerHelper.createConstStatement(varName, def.expression!);
  }

  private getArrayPropertyAsSpread(identifier: ts.Identifier, property: string) {
    return ts.factory.createSpreadElement(
      ts.factory.createPropertyAccessChain(identifier, undefined, property));
  }
}