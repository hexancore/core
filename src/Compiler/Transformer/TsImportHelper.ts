import ts from "typescript";
import { TsTransfromerHelper, type ImportFromMapper } from "./TsTransformerHelper";

export class TsImportHelper {

  protected importedIdentifierMapper: (importDec: ts.ImportDeclaration, name: string, context: ts.TransformationContext) => ts.PropertyAccessChain | ts.Identifier;

  public constructor(
    protected importFromMapper: ImportFromMapper,
    protected needFixImportAccess: boolean
  ) {
    this.importedIdentifierMapper = this.needFixImportAccess ? (importDec, name, context) =>
      TsTransfromerHelper.createNamedImportAccess(importDec, name, context)
      : (_importDec, name, _context) => ts.factory.createIdentifier(name);
  }

  public mapFrom(importFrom: string): string {
    return this.importFromMapper(importFrom);
  }

  public mapAccess(importDec: ts.ImportDeclaration, name: string, context: ts.TransformationContext): ts.PropertyAccessChain | ts.Identifier {
    return this.importedIdentifierMapper(importDec, name, context);
  }
}