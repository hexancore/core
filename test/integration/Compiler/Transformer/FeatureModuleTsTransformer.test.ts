/**
 * @group integration
 */

import { TsTransformerTestHelper } from "@/Compiler/Test/TsTransformerTestHelper";
import { FeatureModuleTsTransformer } from "@/Compiler/Transformer/FeatureModuleTsTransformer";
import { readFileSync } from "fs";
import ts from "typescript";

describe(FeatureModuleTsTransformer.constructor.name, () => {
  const helper = TsTransformerTestHelper.createFromTsConfig(process.cwd() + "/tsconfig.json");
  let transformer: FeatureModuleTsTransformer;

  beforeAll(() => {
    const sourceRoot = process.cwd() + "/test/helper/libs/test-lib/src";
    const compilerRoot = process.cwd() + "/lib/Compiler";
    transformer = FeatureModuleTsTransformer.create(sourceRoot, compilerRoot);
  });

  test("transform", () => {
    const codeFilePath = process.cwd() + "/test/helper/libs/test-lib/src/Book/BookModule.ts";
    const sourceFile = helper.createSourceFileFromExisting(codeFilePath);
    const tsConfig = JSON.parse(readFileSync(process.cwd() +"/tsconfig.json").toString());

    const out = ts.transpileModule(sourceFile.text, {compilerOptions: tsConfig.compilerOptions, transformers: {
      before:  [(context) => (sourceFile: ts.SourceFile) => transformer.transform("Book", sourceFile, context)]
    }});

    expect(out.outputText).toMatchSnapshot();
  });

});


