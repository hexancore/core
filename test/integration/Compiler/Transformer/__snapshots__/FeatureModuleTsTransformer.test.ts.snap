// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function transform 1`] = `
""use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BookModule = exports.SOME_ASSIGMENT_NOISE = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const BookCreateCommandHandler_1 = require("./Application/Book/Command/Create/BookCreateCommandHandler");
const BookGetByIdQueryHandler_1 = require("./Application/Book/Query/GetById/BookGetByIdQueryHandler");
const BookService_1 = require("./Application/Book/Service/BookService");
const BookInfrastructureModule_1 = require("./Infrastructure/BookInfrastructureModule");
// TEST TRANSFORMING IGNORE COMMENT
exports.SOME_ASSIGMENT_NOISE = 10;
let BookModule = class BookModule {
};
exports.BookModule = BookModule;
exports.BookModule = BookModule = tslib_1.__decorate([
    (0, common_1.Module)({ imports: [BookInfrastructureModule_1.BookInfrastructureModule], providers: [BookCreateCommandHandler_1.BookCreateCommandHandler, BookGetByIdQueryHandler_1.BookGetByIdQueryHandler, BookService_1.BookService] })
], BookModule);
"
`;
